gtda/images/filtrations.py

```diff
@@ -1036,3 +1028,227 @@ def plot(Xt, sample=0, colorscale='greys', origin='upper',
+        "euclidean", "manhattan", or "cosine".
+        If ``metric`` is a callable function, it is called on each pair of
+        instances and the resulting value recorded. The callable should take
+        two arrays from the entry in `X` as input, and return a value
+        indicating the distance between them.
+
+    metric_params : dict, optional, default: ``{}``
+        Additional keyword arguments for the metric function.
+
+    n_jobs : int or None, optional, default: ``None``
+        The number of jobs to use for the computation. ``None`` means 1 unless
+        in a :obj:`joblib.parallel_backend` context. ``-1`` means using all
+        processors.
+
+    Attributes
+    ----------
+    mask_ : ndarray of shape (radius, radius, [, radius])
+        The mask applied around each pixel to calculate the weighted number of
+        its activated neighbors. Set in :meth:`fit`.
+
+    See also
+    --------
+    gtda.homology.CubicalPersistence, Binarizer
+
+    References
+    ----------
+    [1] A. Garin and G. Tauzin, "A topological reading lesson: Classification
+        of MNIST  using  TDA"; 19th International IEEE Conference on Machine
+        Learning and Applications (ICMLA 2020), 2019; arXiv: `1910.08345 \
+        <https://arxiv.org/abs/1910.08345>`_.
+
+    """
+
+    _hyperparameters = {
+        'radius': {'type': Real, 'in': Interval(0, np.inf, closed='right')},
+        'metric': {'type': (str, FunctionType)},
+        'metric_params': {'type': dict},
+    }
+
+    def __init__(self, radius=3, metric='euclidean', metric_params={},
+                 n_jobs=None):
+        self.radius = radius
+        self.metric = metric
+        self.metric_params = metric_params
+        self.n_jobs = n_jobs
+
+    def _calculate_density(self, X):
+        Xd = np.zeros(X.shape)
+
+        for i, j, k in self._iterator:
+            Xd += np.roll(np.roll(
+                np.roll(X, k, axis=3), j, axis=2), i, axis=1) \
+                * self.mask_[self._range + i, self._range + j,
+                             self._range + k]
+        return Xd
+
+    def fit(self, X, y=None):
+        """Calculate :attr:`mask_` from a collection of binary images. Then,
+        return the estimator.
+
+        This method is here to implement the usual scikit-learn API and hence
+        work in pipelines.
+
+        Parameters
+        ----------
+        X : ndarray of shape (n_samples, n_pixels_x, n_pixels_y [, n_pixels_z])
+            Input data. Each entry along axis 0 is interpreted as a 2D or 3D
+            binary image.
+
+        y : None
+            There is no need of a target in a transformer, yet the pipeline API
+            requires this parameter.
+
+        Returns
+        -------
+        self : object
+
+        """
+        X = check_array(X, allow_nd=True)
+        self._n_dimensions = X.ndim - 1
+        if (self._n_dimensions < 2) or (self._n_dimensions > 3):
```

Comment:
I'm guessing that, unlike some of the other transformers in `gtda.images`, this one will break on arrays which are not 3D. In your opinion, is it better to extend support to n-dimensional images, or turn this warning into an exception?

Reply:
Yes, that's a tradeoff. I don't think I want to work on extending the transformers to images of arbitrary dimensions simply because I am not sure anyone you would use it.

At the same time, I won't deny them the possibility of trying to make it work. Let's fix this in October?

Reply:
Sure! Nonetheless, why not throwing exceptions for now, instead of warnings?

Reply:
If you throw exceptions, noone will be able to try passing 4d images, no?

Reply:
Right, but this transformer will break there (many `3`s hardcoded in the code), unless it is suitably generalized.

Reply:
Yeah, you're right! I changed it!

Reply:
Maybe I misunderstood, but I still see a lot of `3`s in the code.

Reply:
The mask it computes is 3D, that's why images are reshaped in `transform`. I can then simply use `np.roll` in `_calculate_density`.